<a href="https://colab.research.google.com/github/r-ap/TrainingMachineLearning/blob/main/Assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src = "https://global-uploads.webflow.com/5f0d53c042a9ed6288de7f8d/5f6337ae2cfaa10946ceeb06_Hacktiv8%20logo%20horizontal%2001%20black-p-500.png" width = 400>
<h1 align=center><font size = 5>Hacktiv8 PTP Introduction to Data Science Projects 2 // Statistical Treatment for Datasets</font></h1>

Title: Hacktiv8 PTP Introduction to Data Science Projects 2: Statistical Treatment for Datasets Starter Notebook\
Last Updated: September 20, 2020\
Author: Raka Ardhi

## NYC Property Sales Introduction

The aim of this projects is to introduce you to practical statistic with Python as concrete and as consistent as possible. Using what you’ve learned; download the NYC Property Sales Dataset from Kaggle. This dataset is a record of every building or building unit (apartment, etc.) sold in the New York City property market over a 12-month period.

This dataset contains the location, address, type, sale price, and sale date of building units sold. A reference on the trickier fields:

* `BOROUGH`: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).
* `BLOCK`; `LOT`: The combination of borough, block, and lot forms a unique key for property in New York City. Commonly called a BBL.
* `BUILDING CLASS AT PRESENT` and `BUILDING CLASS AT TIME OF SALE`: The type of building at various points in time.

Note that because this is a financial transaction dataset, there are some points that need to be kept in mind:

* Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement.
* This dataset uses the financial definition of a building/building unit, for tax purposes. In case a single entity owns the building in question, a sale covers the value of the entire building. In case a building is owned piecemeal by its residents (a condominium), a sale refers to a single apartment (or group of apartments) owned by some individual.

Formulate a question and derive a statistical hypothesis test to answer the question. You have to demonstrate that you’re able to make decisions using data in a scientific manner. Examples of questions can be:

* Is there a difference in unit sold between property built in 1900-2000 and 2001 so on?
* Is there a difference in unit sold based on building category?
* What can you discover about New York City real estate by looking at a year's worth of raw transaction records? Can you spot trends in the market?

Please make sure that you have completed the lesson for this course, namely Python and Practical Statistics which is part of this Program.

**Note:** You can take a look at Project Rubric below:

| Code Review |  |
| :--- | :--- |
| CRITERIA | SPECIFICATIONS |
| Mean | Student implement mean to specifics column/data using pandas, numpy, or scipy|
| Median | Student implement median to specifics column/data using pandas, numpy, or scipy|
| Modus | Student implement modus to specifics column/data using pandas, numpy, or scipy|
| Central Tendencies | Implementing Central Tendencies through dataset |
| Box Plot | Implementing Box Plot to visualize spesific data |
| Z-Score | Implementing Z-score concept to specific data |
| Probability Distribution | Student analyzing distribution of data and gain insight from the distribution |
| Intervals | Implementing Confidence or Prediction Intervals |
| Hypotesis Testing | Made 1 Hypotesis and get conclusion from data |
| Preprocessing | Student preprocess dataset before applying the statistical treatment. |
| Does the code run without errors? | The code runs without errors. All code is functional and formatted properly. |

| Readability |  |
| :--- | :--- |
| CRITERIA | SPECIFICATIONS |
| Well Documented | All cell in notebook are well documented with markdown above each cell explaining the code|

| Analysis |  |
| :--- | :--- |
| CRITERIA | SPECIFICATIONS |
|Overall Analysis| Gain an insight/conclusion of overall plots that answer the hypotesis |

**Focus on "Graded-Function" sections.**

------------

## Data Preparation

Load the library you need.

Get your NYC property data from [here](https://www.kaggle.com/new-york-city/nyc-property-sales) and load the dataframe to your notebook.

In [ ]:
# Get your import statement here
import numpy as np
import pandas as pd


In [ ]:
# Load your dataset here
from google.colab import files
uploaded = files.upload()
print ('Data read into a pandas dataframe!')

Saving nyc-rolling-sales.csv to nyc-rolling-sales (1).csv
Data read into a pandas dataframe!


In [ ]:
import io
df = pd.read_csv(io.BytesIO(uploaded['nyc-rolling-sales.csv']))

Let's view the top 5 rows of the dataset using the `head()` function.

In [ ]:
# Write your syntax here
df.head()

Unnamed: 0  BOROUGH   NEIGHBORHOOD  \
0           4        1  ALPHABET CITY   
1           5        1  ALPHABET CITY   
2           6        1  ALPHABET CITY   
3           7        1  ALPHABET CITY   
4           8        1  ALPHABET CITY   

                       BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  BLOCK  \
0  07 RENTALS - WALKUP APARTMENTS                                2A    392   
1  07 RENTALS - WALKUP APARTMENTS                                 2    399   
2  07 RENTALS - WALKUP APARTMENTS                                 2    399   
3  07 RENTALS - WALKUP APARTMENTS                                2B    402   
4  07 RENTALS - WALKUP APARTMENTS                                2A    404   

   LOT EASE-MENT BUILDING CLASS AT PRESENT                 ADDRESS  ...  \
0    6                                  C2            153 AVENUE B  ...   
1   26                                  C7   234 EAST 4TH   STREET  ...   
2   39                                  C7   197 EAST 3RD   STREET  ...   
3   21                                  C4     154 EAST 7TH STREET  ...   
4   55                                  C2  301 EAST 10TH   STREET  ...   

  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
0                 5                 0            5              1633   
1                28                 3           31              4616   
2                16                 1           17              2212   
3                10                 0           10              2272   
4                 6                 0            6              2369   

   GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
0               6440       1900                         2   
1              18690       1900                         2   
2               7803       1900                         2   
3               6794       1913                         2   
4               4615       1900                         2   

   BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
0                              C2     6625000  2017-07-19 00:00:00  
1                              C7         -    2016-12-14 00:00:00  
2                              C7         -    2016-12-09 00:00:00  
3                              C4     3936272  2016-09-23 00:00:00  
4                              C2     8000000  2016-11-17 00:00:00  

[5 rows x 22 columns]

We can also veiw the bottom 5 rows of the dataset using the `tail()` function.

In [ ]:
# Write your syntax here
df.tail()

Unnamed: 0  BOROUGH NEIGHBORHOOD  \
84543        8409        5      WOODROW   
84544        8410        5      WOODROW   
84545        8411        5      WOODROW   
84546        8412        5      WOODROW   
84547        8413        5      WOODROW   

                           BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  \
84543  02 TWO FAMILY DWELLINGS                                        1   
84544  02 TWO FAMILY DWELLINGS                                        1   
84545  02 TWO FAMILY DWELLINGS                                        1   
84546  22 STORE BUILDINGS                                             4   
84547  35 INDOOR PUBLIC AND CULTURAL FACILITIES                       4   

       BLOCK  LOT EASE-MENT BUILDING CLASS AT PRESENT                ADDRESS  \
84543   7349   34                                  B9          37 QUAIL LANE   
84544   7349   78                                  B9       32 PHEASANT LANE   
84545   7351   60                                  B2       49 PITNEY AVENUE   
84546   7100   28                                  K6  2730 ARTHUR KILL ROAD   
84547   7105  679                                  P9      155 CLAY PIT ROAD   

       ... RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
84543  ...                 2                 0            2              2400   
84544  ...                 2                 0            2              2498   
84545  ...                 2                 0            2              4000   
84546  ...                 0                 7            7            208033   
84547  ...                 0                 1            1             10796   

       GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
84543               2575       1998                         1   
84544               2377       1998                         1   
84545               1496       1925                         1   
84546              64117       2001                         4   
84547               2400       2006                         4   

       BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
84543                              B9      450000  2016-11-28 00:00:00  
84544                              B9      550000  2017-04-21 00:00:00  
84545                              B2      460000  2017-07-05 00:00:00  
84546                              K6    11693337  2016-12-21 00:00:00  
84547                              P9       69300  2016-10-27 00:00:00  

[5 rows x 22 columns]

BOROUGH: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).

To view the dimensions of the dataframe, we use the `.shape` parameter. Expected result: (84548, 22)

In [ ]:
# Write your syntax here
df.shape

(84548, 22)

According to this official page, Ease-ment is "is a right, such as a right of way, which allows an entity to make limited use of another’s real property. For example: MTA railroad tracks that run across a portion of another property". Also, the Unnamed column is not mentioned and was likely used for iterating through records. So, those two columns are removed for now.

In [ ]:
# Drop 'Unnamed: 0' and 'EASE-MENT' features using .drop function
df = df.drop(columns=['Unnamed: 0','EASE-MENT'])


Let's view Dtype of each features in dataframe using `.info()` function.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   BOROUGH                         84548 non-null  int64 
 1   NEIGHBORHOOD                    84548 non-null  object
 2   BUILDING CLASS CATEGORY         84548 non-null  object
 3   TAX CLASS AT PRESENT            84548 non-null  object
 4   BLOCK                           84548 non-null  int64 
 5   LOT                             84548 non-null  int64 
 6   BUILDING CLASS AT PRESENT       84548 non-null  object
 7   ADDRESS                         84548 non-null  object
 8   APARTMENT NUMBER                84548 non-null  object
 9   ZIP CODE                        84548 non-null  int64 
 10  RESIDENTIAL UNITS               84548 non-null  int64 
 11  COMMERCIAL UNITS                84548 non-null  int64 
 12  TOTAL UNITS                     84548 non-null

It looks like empty records are not being treated as NA. We convert columns to their appropriate data types to obtain NAs.

In [ ]:
#First, let's check which columns should be categorical
print('Column name')
for col in df.columns:
    if df[col].dtype=='object':
        print(col, df[col].nunique())

Column name
NEIGHBORHOOD 254
BUILDING CLASS CATEGORY 47
TAX CLASS AT PRESENT 11
BUILDING CLASS AT PRESENT 167
ADDRESS 67563
APARTMENT NUMBER 3989
LAND SQUARE FEET 6062
GROSS SQUARE FEET 5691
BUILDING CLASS AT TIME OF SALE 166
SALE PRICE 10008
SALE DATE 364


In [ ]:
# LAND SQUARE FEET,GROSS SQUARE FEET, SALE PRICE, BOROUGH should be numeric. 
# SALE DATE datetime format.
# categorical: NEIGHBORHOOD, BUILDING CLASS CATEGORY, TAX CLASS AT PRESENT, BUILDING CLASS AT PRESENT,
# BUILDING CLASS AT TIME OF SALE, TAX CLASS AT TIME OF SALE,BOROUGH 

numer = ['LAND SQUARE FEET','GROSS SQUARE FEET', 'SALE PRICE', 'BOROUGH']
for col in numer: # coerce for missing values
    df[col] = pd.to_numeric(df[col], errors='coerce')

categ = ['NEIGHBORHOOD', 'BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT', 'BUILDING CLASS AT PRESENT', 'BUILDING CLASS AT TIME OF SALE', 'TAX CLASS AT TIME OF SALE']
for col in categ:
    df[col] = df[col].astype('category')

df['SALE DATE'] = pd.to_datetime(df['SALE DATE'], errors='coerce')

Our dataset is ready for checking missing values.

In [ ]:
missing = df.isnull().sum()/len(df)*100

print(pd.DataFrame([missing[missing>0],pd.Series(df.isnull().sum()[df.isnull().sum()>1000])], index=['percent missing','how many missing']))

                  LAND SQUARE FEET  GROSS SQUARE FEET   SALE PRICE
percent missing          31.049818          32.658372     17.22217
how many missing      26252.000000       27612.000000  14561.00000


Around 30% of GROSS SF and LAND SF are missing. Furthermore, around 17% of SALE PRICE is also missing.

We can fill in the missing value from one column to another, which will help us reduce missing values. Expected values:

(6, 20)

(1366, 20)

In [ ]:
print(df[(df['LAND SQUARE FEET'].isnull()) & (df['GROSS SQUARE FEET'].notnull())].shape)
print(df[(df['LAND SQUARE FEET'].notnull()) & (df['GROSS SQUARE FEET'].isnull())].shape)

(6, 20)
(1366, 20)


There are 1372 rows that can be filled in with their approximate values.

In [ ]:
df['LAND SQUARE FEET'] = df['LAND SQUARE FEET'].mask((df['LAND SQUARE FEET'].isnull()) & (df['GROSS SQUARE FEET'].notnull()), df['GROSS SQUARE FEET'])
df['GROSS SQUARE FEET'] = df['GROSS SQUARE FEET'].mask((df['LAND SQUARE FEET'].notnull()) & (df['GROSS SQUARE FEET'].isnull()), df['LAND SQUARE FEET'])

In [ ]:
#  Check for duplicates before

print(sum(df.duplicated()))

df[df.duplicated(keep=False)].sort_values(['NEIGHBORHOOD', 'ADDRESS']).head(10)

# df.duplicated() automatically excludes duplicates, to keep duplicates in df we use keep=False

# in df.duplicated(df.columns) we can specify column names to look for duplicates only in those mentioned columns.

0


Empty DataFrame
Columns: [Unnamed: 0, BOROUGH, NEIGHBORHOOD, BUILDING CLASS CATEGORY, TAX CLASS AT PRESENT, BLOCK, LOT, EASE-MENT, BUILDING CLASS AT PRESENT, ADDRESS, APARTMENT NUMBER, ZIP CODE, RESIDENTIAL UNITS, COMMERCIAL UNITS, TOTAL UNITS, LAND SQUARE FEET, GROSS SQUARE FEET, YEAR BUILT, TAX CLASS AT TIME OF SALE, BUILDING CLASS AT TIME OF SALE, SALE PRICE, SALE DATE]
Index: []

[0 rows x 22 columns]

The dataframe has 765 duplicated rows (exluding the original rows).

In [ ]:
df.drop_duplicates(inplace=True)

print(sum(df.duplicated()))

0


## Exploratory data analysis

Now, let's get a simple descriptive statistics with `.describe()` function for `COMMERCIAL UNITS` features.

In [ ]:
df[df['COMMERCIAL UNITS']==0].describe()

BOROUGH         BLOCK           LOT      ZIP CODE  \
count  78777.000000  78777.000000  78777.000000  78777.000000   
mean       3.004329   4273.781015    395.422420  10722.737068   
std        1.298594   3589.241940    671.604654   1318.493961   
min        1.000000      1.000000      1.000000      0.000000   
25%        2.000000   1330.000000     23.000000  10304.000000   
50%        3.000000   3340.000000     52.000000  11209.000000   
75%        4.000000   6361.000000   1003.000000  11357.000000   
max        5.000000  16322.000000   9106.000000  11694.000000   

       RESIDENTIAL UNITS  COMMERCIAL UNITS   TOTAL UNITS  LAND SQUARE FEET  \
count       78777.000000           78777.0  78777.000000      5.278000e+04   
mean            1.691737               0.0      1.724133      3.140140e+03   
std             9.838994               0.0      9.835016      2.929999e+04   
min             0.000000               0.0      0.000000      0.000000e+00   
25%             0.000000               0.0      1.000000      1.600000e+03   
50%             1.000000               0.0      1.000000      2.295000e+03   
75%             2.000000               0.0      2.000000      3.300000e+03   
max           889.000000               0.0    889.000000      4.252327e+06   

       GROSS SQUARE FEET    YEAR BUILT    SALE PRICE  
count       5.278000e+04  78777.000000  6.562900e+04  
mean        2.714612e+03   1781.065451  9.952969e+05  
std         2.791294e+04    551.024570  3.329268e+06  
min         0.000000e+00      0.000000  0.000000e+00  
25%         9.750000e+02   1920.000000  2.400000e+05  
50%         1.600000e+03   1940.000000  5.294900e+05  
75%         2.388000e+03   1967.000000  9.219560e+05  
max         4.252327e+06   2017.000000  3.450000e+08

Let us try to understand the columns. Above table shows descriptive statistics for the numeric columns.

- There are zipcodes with 0 value
- Can block/lot numbers go up to 16322?
- Most of the properties have 2 unit and maximum of 1844 units? The latter might mean some company purchased a building. This should be treated as an outlier.
- Other columns also have outliers which needs further investigation.
- Year column has a year with 0
- Most sales prices less than 10000 can be treated as gift or transfer fees.

Now, let's get a simple descriptive statistics with `.describe()` function for `RESIDENTIAL UNITS` features.

Function below are graded function. (1 Points)

In [ ]:
df[df['RESIDENTIAL UNITS']==0].describe()

BOROUGH         BLOCK           LOT      ZIP CODE  \
count  24546.000000  24546.000000  24546.000000  24546.000000   
mean       2.542084   3355.267172    283.943412  10321.514626   
std        1.334486   3091.222257    570.045258   2135.406385   
min        1.000000      1.000000      1.000000      0.000000   
25%        1.000000   1158.000000     12.000000  10023.000000   
50%        3.000000   1947.000000     38.000000  11004.000000   
75%        4.000000   5390.750000    135.000000  11354.000000   
max        5.000000  16317.000000   9056.000000  11694.000000   

       RESIDENTIAL UNITS  COMMERCIAL UNITS   TOTAL UNITS  LAND SQUARE FEET  \
count            24546.0      24546.000000  24546.000000      9.503000e+03   
mean                 0.0          0.459382      0.563350      7.416797e+03   
std                  0.0         15.826020     15.825945      8.032892e+04   
min                  0.0          0.000000      0.000000      0.000000e+00   
25%                  0.0          0.000000      0.000000      0.000000e+00   
50%                  0.0          0.000000      0.000000      0.000000e+00   
75%                  0.0          0.000000      0.000000      3.250000e+03   
max                  0.0       2261.000000   2261.000000      4.252327e+06   

       GROSS SQUARE FEET    YEAR BUILT    SALE PRICE  
count       9.503000e+03  24546.000000  2.085500e+04  
mean        8.870466e+03   1675.525788  1.632257e+06  
std         7.890877e+04    679.095000  1.969307e+07  
min         0.000000e+00      0.000000  0.000000e+00  
25%         0.000000e+00   1921.000000  1.825000e+05  
50%         0.000000e+00   1950.000000  3.950000e+05  
75%         2.500000e+03   1962.000000  8.500000e+05  
max         4.252327e+06   2017.000000  2.210000e+09

Write your findings below:

Use `.value_counts` function to count total value of `BOROUGH` features. Expected value:

4    26548\
3    23843\
1    18102\
5     8296\
2     6994\
Name: BOROUGH, dtype: int64

In [ ]:
df["BOROUGH"].value_counts()

4    26548
3    23843
1    18102
5     8296
2     6994
Name: BOROUGH, dtype: int64

From here, we can calculate the mean for each Borough. Use `.mean()` function to calculate mean.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin (~1 Lines)

# Graded-Fu

mean = df_BOROUGH['nyc-rolling-sales.csv'].mean()

NameError: ignored

From here, we can calculate the median for each Borough. Use `.median()` function to calculate median.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

From here, we can calculate the mode for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

From here, we can calculate the Range for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

From here, we can calculate the Variance for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

From here, we can calculate the SD for each Borough.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin (~1 Lines)

# Graded-Funtion End

Now we can analyze Probability Distibution below.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin





# Graded-Funtion End

Now we can analyze Confidence Intervals below.

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin





# Graded-Funtion End

Make your Hypothesis Testing below

Function below are graded function. (1 Points)

In [ ]:
# Write your function below

# Graded-Funtion Begin





# Graded-Funtion End

Write your final conclusion below.

Your conclusion below are graded. (1 Points)